### PR TITLE
Instant download gene

### PR DESCRIPTION
**Pull request by imransl**
## Type

- [ ] Bug fix
- [x] New feature
- [ ] Improvement to existing feature

## Related JIRA Issue(s)
[ENSWBSITES-862](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-862)
[ENSWBSITES-966](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-966)
[ENSWBSITES-785](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-785)

## Description
This PR adds a new component for the instant gene download. It is added into the drawer in the genome browser. The genomic sequence is retrieved via REST as it is still not available via RefGet. However, the other sequences are retrieved from the latter.

Also we now display stable ids in Transcript Instant Download too.

I have added the component in genome browser drawer for which the XD is: https://xd.adobe.com/view/f4540088-a666-4eaa-5cae-dd51a317e4d4-8265/screen/63ab6e61-e39b-4f6f-978d-2078034324cc/?fullscreen.

Also I have added it to the entity viewer sidebar. The XD for it is: https://xd.adobe.com/view/75304d3a-0f51-428d-9f47-bca20ee7201a-2063/screen/ab07c6ff-fe41-4dc7-868d-d3af7c0a32bb/?fullscreen

## Deployment URL
http://instant-download-gene.review.ensembl.org/

## Views affected
Genome browser -> track panel -> drawer for gene

### Other effects
- [ ] I have checked any requirements for Google Analytics
- [ ] I have created any required tests
- [ ] The PR has as its final commit a WASM/JS update commit if any Rust files were updated.

### CODE REVIEW

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.tsx [hunk 1]
```diff
@@ -122,27 +128,50 @@ const GeneSummary = () => {
...
+            className={styles.downloadLink}
+            onClick={() => showDownload(!shouldShowDownload)}
+          >
+            Download
+          </div>
+          {shouldShowDownload && (
+            <div className={styles.downloadWrapper}>
+              <InstantDownloadGene
+                genomeId={ensObjectGene.genome_id}
+                gene={{ id: gene.unversioned_stable_id }}
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.tsx:163]: In https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-785 (I don't think he's pushed his changes to github yet) @jyothishnt is adding versions to stable ids both in the UI and in the headings of the downloaded sequence. Which means that he is passing a regular `stable_id` to the instant download component, and then does various tricks to fetch the unversioned stable id from thoas that he will need for fetching genomic sequences from REST.

I wonder if the approach in both PRs should be similar.

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/EntityViewerSidebarModal.tsx [hunk 2]
```diff
@@ -36,6 +36,9 @@ const entityViewerSidebarModals: Record<
   [SidebarModalView.BOOKMARKS]: lazy(
     () => import('./modal-views/EntityViewerBookmarks')
   ),
+  [SidebarModalView.SHARE]: lazy(
+    () => import('./modal-views/EntityViewerShare')
+  ),
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/EntityViewerSidebarModal.tsx:41]: I'm wondering whether we shouldn't just delete `./modal-views/EntityViewerShare` and disable the share button in the toolbar. We aren't going to have a share functionality for the MAP, and maybe not even for the MVP; so we shouldn't really keep those empty components.

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  });
+
+  const sequenceDownloadParams = flattenDeep(
+    geneSequenceData.transcripts.map((transcript) =>
+      prepareDownloadParameters({
+        transcriptSequenceData: transcript,
+        options: transcriptOptions
+      })
+    )
+  );
```
Comment (azangru) [inline src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts:63]: I'm confused: what's being flattened here?

Also, if there's something really being flattened, won't the native Array.prototype.flat or .flatMap do the job?

File: src/ensembl/src/shared/workers/sequenceFetcher.worker.ts [hunk 4]
```diff
@@ -35,7 +35,7 @@ const downloadSequences = async (params: SequenceFetcherParams) => {
   const sequences = await Promise.all(sequencePromises);
 
   // start new sequence on a new line; no empty lines allowed in FASTA files
-  return sequences.join('\n');
+  return sequences.join('\n\n');
```
Comment (azangru) [inline src/ensembl/src/shared/workers/sequenceFetcher.worker.ts:38]: 🙅‍♂️ 

no empty lines in fasta files!

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.tsx (hunk 1 above)
Comment (imransl) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.tsx:163]: Yes agree they should be similar. I'll check him PR and try to replicate that here.

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+import downloadAsFile from 'src/shared/helpers/downloadAsFile';
+
+import { TranscriptOptions } from '../instant-download-transcript/InstantDownloadTranscript';
+import { fetchGeneSequenceMetadata } from './fetchSequenceChecksums';
+
+import { WorkerApi } from 'src/shared/workers/sequenceFetcher.worker';
+import {
+  getGenomicSequenceData,
+  prepareDownloadParameters
+} from './fetchForTranscript';
```
[the remark this answers is not in the document]
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts:28]: As we discussed during the standup, I'll leave the code as it is for now. We are anyway sending only one request to Thoas at the moment as we use a different query and function for it. Strangely `getGenomicSequenceData` (why this name? 😲 ) does not retrieve any data but only creates the url and label that we need to retrieve and process the data. So think we can leave it as it is for now?

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts (hunk 3 above)
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts:61]: I used `flatMap` for this ✅

File: src/ensembl/src/shared/workers/sequenceFetcher.worker.ts (hunk 4 above)
Comment (imransl) [reply src/ensembl/src/shared/workers/sequenceFetcher.worker.ts]: ✅

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.tsx (hunk 1 above)
Comment (imransl) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.tsx]: I am now passing the versioned stabled ID to `InstantDownloadGene`. The unversioned stabled IDs are retrieved inside the component from Thoas to request genomic sequences via REST.

✅

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts:28]: I've pushed a commit which does the following:

- Moves fetching of unversioned_stable_id from the `InstantDownloadGene` component to the functions within the fetchSequenceChecksums file. Reasons: 1) the rendering of the `InstantDownloadGene` component should not depend on the fetching of the unversioned stable id, and 2) we are already fetching other kinds of sequence-related information for the gene and its transcripts in those functions
- Undoes the overwriting of the versioned stable id with the unversioned stable id that was happening inside the `fetchGeneSequenceMetadata` function. We can include both versioned and unversioned stable ids in the value returned from that function; we shouldn't need to overwrite one with other. Also, I am sure we would prefer versioned stable ids in the comments in the downloaded fasta file.

I think these changes should also now cover @jyothishnt's use case for downloading sequences for individual transcripts.

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss [hunk 6]
```diff
@@ -38,10 +37,35 @@
...
+.featureSymbol + .stableId {
   margin-left: 12px;
   font-weight: $light;
 }
+
+.downloadRow {
+  > .value {
+    display: flex;
+    position: relative;
+    width: 400px;
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss:52]: I tried to remove `position:relative` and `width:400px` from the declaration block, and couldn't see any difference in the layout. Are these two rules needed for anything?

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss [hunk 7]
```diff
@@ -38,10 +37,35 @@
...
+  }
+}
+
+.downloadLink {
+  color: $blue;
+  cursor: pointer;
+}
+
+.downloadWrapper {
+  display: inline-block;
```
Comment (azangru) [inline src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss:62]: The inline-block rule doesn't seem to change anything. `display:flex` of the parent is already positioning the two elements next to one another.

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss (hunk 6 above)
Comment (imransl) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss:52]: Think the styling here and in the comment mentioned below are obsolete. I'll remove them.

Comment (imransl) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss]: ✅

File: src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss (hunk 7 above)
Comment (imransl) [reply src/ensembl/src/content/app/browser/drawer/drawer-views/gene-summary/GeneSummary.scss]: ✅

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-toolstrip/EntityViewerSidebarToolstrip.tsx [hunk 8]
```diff
@@ -90,11 +89,10 @@ export const EntityViewerSidebarToolstrip = () => {
         image={shareIcon}
       />
       <ImageButton
-        status={Status.DISABLED}
-        description="Downloads"
+        status={getViewIconStatus(SidebarModalView.DOWNLOADS)}
+        description={SidebarModalView.DOWNLOADS}
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-toolstrip/EntityViewerSidebarToolstrip.tsx:93]: Not important at the moment, but `description` is free text that doesn't need to be captured in enums. It's just a lucky chance that the descriptions for these buttons currently happen to be the same as our internal strings representing their statuses.

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchSequenceChecksums.ts [hunk 9]
```diff
@@ -33,26 +35,56 @@ export type TranscriptSequenceMetadata = {
   };
 };
 
+export type GeneSequenceMetadata = {
+  stable_id: string;
+  unversioned_stable_id: string;
+  transcripts: TranscriptSequenceMetadata[];
+};
+
+type TranscriptInQuery = {
```
Comment (azangru) [inline src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchSequenceChecksums.ts:44]: The name of the type is misleading here. The query is something that we send to the api; while this type is describing what has been returned from the api. Is it `TranscriptInGeneQueryResult`? Or perhaps `TranscriptInResponse`?

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.scss [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+  .featureId {
+    margin-left: 0.5em;
+    font-weight: $bold;
+  }
+}
+
+.transcriptVis {
+  margin-top: 18px;
+  margin-bottom: 9px;
+}
```
Comment (azangru) [inline src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.scss:21]: Could you please make sure that all the classes in this file are actually used in the layout? This class definitely is not.

**Comment by azangru** [comment]
There are several issues that @ens-st3 pointed out in his first comment on March 15 [here](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-966?focusedCommentId=354158&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354158) that need to be fixed before this PR can be merged.

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jyothishnt) [reply src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchForGene.ts:28]: @azangru  yes, it fixes the one I am working on. Though I needed some minor changes which is now pushed into this branch.

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-toolstrip/EntityViewerSidebarToolstrip.tsx (hunk 8 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-toolstrip/EntityViewerSidebarToolstrip.tsx]: That's done now ✅

**Comment by imransl** [comment]
@azangru @ens-st3 I've made the following changes based on the QA comments from Steve:

* cDNA, CDS and protein sequence options are not shown for non-coding genes
* The order of sequences for each transcript now is: genomic sequence, cDNA, CDS and protein sequence
* The tooltip for download icon has been changed to 'Download'

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.scss (hunk 10 above)
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.scss]: ✅

File: src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchSequenceChecksums.ts (hunk 9 above)
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-fetch/fetchSequenceChecksums.ts]: ✅

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+import { fetchForGene } from '../instant-download-fetch/fetchForGene';
+import { filterTranscriptOptions } from '../instant-download-transcript/InstantDownloadTranscript';
+
+import Checkbox from 'src/shared/components/checkbox/Checkbox';
+import InstantDownloadButton from '../instant-download-button/InstantDownloadButton';
+
+import styles from './InstantDownloadGene.scss';
+
+type Theme = 'light' | 'dark';
```
Comment (azangru) [inline src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx:29]: Out of curiosity: does `InstantDownloadGene` even need a theme prop?

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerDownloads.scss [hunk 12]
```diff
@@ -0,0 +1,5 @@
+.container {
+  h3 {
+    margin: 5px 0 25px 0;
+  }
+}
```
Comment (azangru) [inline src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerDownloads.scss:5]: empty line?

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx (hunk 11 above)
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx:29]: I'll check with Andrea about this and see if we need this

File: src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerDownloads.scss (hunk 12 above)
Comment (imransl) [reply src/ensembl/src/content/app/entity-viewer/shared/components/entity-viewer-sidebar/entity-viewer-sidebar-modal/modal-views/EntityViewerDownloads.scss]: ✅

File: src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx (hunk 11 above)
Comment (imransl) [reply src/ensembl/src/shared/components/instant-download/instant-download-gene/InstantDownloadGene.tsx:29]: I asked Andrea about this. She said that we need the dark theme for the Genome browser zmenus:

```
Andrea Winterbottom  14:22
yes - but the download in the genome browser image should allow users to get the same data - however
at zoom where individual transcripts are being rendered, the download component in the zmenus should be the same as the transcript info panel in EV
and at zoom where the gene is being rendered as a block without the ability to identify individual transcripts, then only the gene download option should be available
```

**Review by azangru — APPROVED** [review]
Ok, I guess it's good to go. Don't know if you want Steve to give it a final check.

**Comment by imransl** [comment]
> Ok, I guess it's good to go. Don't know if you want Steve to give it a final check.

I think Steve is fine with merging as he said it's alright to do it, during the last week's sprint review.

**Pull request merged**